From 3489c486f9b737381886ea95c2afd768a58c7943 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:17:43 -0400
Subject: [net] netfilter: nf_tables: fix tracing of the goto action

Message-id: <20141021161742.22888.58728.stgit@localhost.localdomain>
Patchwork-id: 97942
O-Subject: [RHEL7 net PATCH 045/124] netfilter: nf_tables: fix tracing of the goto action
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 7b9d5ef932297413adcbd8be98fe612b9527a312

 commit 7b9d5ef932297413adcbd8be98fe612b9527a312
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Sat May 10 18:42:57 2014 +0200

    netfilter: nf_tables: fix tracing of the goto action

    Add missing code to trace goto actions.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_core.c b/net/netfilter/nf_tables_core.c
index 68d9fda..f3e3892 100644
--- a/net/netfilter/nf_tables_core.c
+++ b/net/netfilter/nf_tables_core.c
@@ -171,8 +171,12 @@ next_rule:
   jumpstack[stackptr].rule  = rule;
   jumpstack[stackptr].rulenum = rulenum;
   stackptr++;
-  /* fall through */
+  chain = data[NFT_REG_VERDICT].chain;
+  goto do_chain;
  case NFT_GOTO:
+  if (unlikely(pkt->skb->nf_trace))
+   nft_trace_packet(pkt, chain, rulenum, NFT_TRACE_RULE);
+
   chain = data[NFT_REG_VERDICT].chain;
   goto do_chain;
  case NFT_RETURN:
-- 
1.7.1